File: src/main/java/spring/webmvc/domain/model/entity/OrderItem.java
package spring.webmvc.domain.model.entity;

import jakarta.persistence.Column;
import jakarta.persistence.ConstraintMode;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@Table(name = "order_item")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class OrderItem {

	@Id
	@GeneratedValue
	@Column(name = "order_item_id")
	private Long id;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "item_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
	private Item item;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "order_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
	private Order order;

	private int orderPrice;
	private int count;

	public static OrderItem create(Item item, int count) {
		OrderItem orderItem = new OrderItem();

		orderItem.item = item;
		orderItem.orderPrice = item.getPrice();
		orderItem.count = count;

		item.removeQuantity(count);

		return orderItem;
	}

	public void associateOrder(Order order) {
		this.order = order;
	}

	public void cancel() {
		item.addQuantity(count);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/model/entity/OrderItem.kt
package spring.web.kotlin.domain.model.entity

import jakarta.persistence.*

@Entity
@Table(name = "order_item")
class OrderItem protected constructor(
    val orderPrice: Int,
    val count: Int,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id", foreignKey = ForeignKey(ConstraintMode.NO_CONSTRAINT))
    var item: Item,
) {
    @Id
    @GeneratedValue
    @Column(name = "order_item_id")
    var id: Long? = null
        protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id", foreignKey = ForeignKey(ConstraintMode.NO_CONSTRAINT))
    var order: Order? = null
        protected set

    companion object {
        fun create(item: Item, count: Int) =
            OrderItem(
                orderPrice = item.price,
                count = count,
                item = item
            ).apply { item.removeQuantity(count) }
    }

    fun associateOrder(order: Order) {
        this.order = order
    }

    fun cancel() {
        item.addQuantity(count)
    }
}